File: HW1/task.java
/**
 * Сортировка кучей, пирамидальная сортировка
 */

 public class HeapSort {

    /**
     * Метод запуска сортировки массива
     * @param args
     */
    public static void main(String[] args) {
      int[] arr = {1234, 4, 7, 2, 1, -3, 0, 567, 45, 90, 34,2, 234};
  
      heapSort(arr);
  
      for (int i = 0; i < arr.length; i++)
        System.out.print(arr[i] + " ");
    }
  
    /**
     * Метод создания дерева массива
     * @param arr
     */
    public static void heapSort(int[] arr) {
      //длина массива
      int n = arr.length;
  
      //создаём дерево, построение кучи (перегруппируем массив)
      for(int i  = n / 2 - 1; i >= 0; i--)
        heapify(arr, i , n);
  
      //Делаем сортировку массива, уже отсортированного дерева,
      //Один за другим извлекаем элементы из кучи
      for (int i = n - 1; i >= 0; i--){
  
        // Перемещаем текущий корень в конец
        int temp = arr[i];
        arr[i] = arr[0];
        arr[0] = temp;
  
        // Вызываем процедуру heapify на уменьшенной куче
        heapify(arr, 0, i);
      }
  
    }
}

File: HW1/task.py
def heapify(arr, n, i, reverse=False):
    largest = i
    while True:
        left = 2 * i + 1
        right = 2 * i + 2

        if left < n and ((arr[i] < arr[left]) if reverse else (arr[i] > arr[left])):
            largest = left

        if right < n and ((arr[largest] < arr[right]) if reverse else (arr[largest] > arr[right])):
            largest = right

        if largest != i:
            arr[i], arr[largest] = arr[largest], arr[i]
            i = largest
        else:
            break

def heap_sort(arr, reverse=False):
    n = len(arr)

    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i, reverse)

    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]
        heapify(arr, i, 0, reverse)

arr = [12, 11, 13, 5, 6, 7, 30, 1, 4, 8]
print("Исходный массив:", arr)

heap_sort(arr)
print("Отсортированный массив (убывание):", arr)

heap_sort(arr, reverse=True)
print("Отсортированный массив (возрастание):", arr)
